### ISSUE TITLE
MemorySanitizer false positive on function call

### ISSUE THREAD

**Issue by ccotter** [llvm/llvm-project#101036]
MSAN appears to report a false positive when passing an uninitialized value to a function that's defined in another translation unit.

```
// other.cpp
void call_other_TU(int) {}

// main.cpp
void call_same_TU(int){}
void call_other_TU(int);

int main() {
    int a;
    call_same_TU(a);
    call_other_TU(a);
    return 0;
}
```

```
Program returned: 1
==1==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55555561ede5 in main /app/main.cpp:7:5
    #1 0x7ffff7829d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 490fef8403240c91833978d494d39e537409b92e)
    #2 0x7ffff7829e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: 490fef8403240c91833978d494d39e537409b92e)
    #3 0x5555555862e4 in _start (/app/build/the_executable+0x322e4)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /app/main.cpp:7:5 in main
Exiting
```

I'm not sure if this is a known limitation (could not find anything in the docs) or intentionally flagged, as the use case seems relatively mundane. As far as I can tell, the program behavior is not observably impacted by the uninitialized value.

https://godbolt.org/z/oGdxhj9Yj

**Comment by vitalybuka** [comment]
with `-fno-inline` it should report `call_same_TU`
We just inline before msan.

Both cases are passing value as argument is "use of uninitialized value", so UB.

**Comment by vitalybuka** [comment]
This is check can be disabled with -fno-sanitize-memory-param-retval

**Comment by ccotter** [comment]
Thanks, `-fno-inline` would explain it.

Can I ask philosophically why by default the behavior appears to be to treat copies of uninitialized memory (UM) to argument is flagged, but copying to a local variable is not? The original MSAN paper states

 > A subset of operations on application values require their operands to be initialized. These operations are, at minimum: conditional branch, system call and pointer dereference. An attempt to pass uninitialized memory as an argument of one of these instructions is reported as an error.

Although this only covers a subset of the situations an error is reported, it seems odd that initialized a function argument with UM is reported by default, but initializing a local stack variable with UM is not.

In either case, it doesn't seem to be a false positive, so thanks for clarifying that!

**Comment by ccotter** [comment]
Ah, https://github.com/google/sanitizers/issues/1755 answers this. This can be closed!
